File: PAT_Basic_Level_by_r/1103 缘分数/1103 缘分数.R
# 题目名称：缘分数
# 题目编号：PAT Basic 1103
# 实现原理：
#   判断区间[m,n]内的数a是否是缘分数。缘分数定义：a和a-1的立方差等于c²，
#   且c等于b和b-1的平方和。解题思路：1. 遍历区间内每个a；
#   2. 计算c² = a³ - (a-1)³，判断是否为完全平方数；
#   3. 若是完全平方数，求c并枚举b，判断b² + (b-1)² = c是否成立；
#   4. 收集所有缘分数对，按格式输出。
# 时间复杂度：O((n-m) × √c)，外层遍历区间，内层枚举b
# 空间复杂度：O(k)，k为缘分数对的数量，用于存储结果
z<-as.integer(scan("stdin",quiet=TRUE));out<-character();for(a in z[1]:z[2]){c2<-a^3-(a-1)^3;c<-round(sqrt(c2));if(c*c==c2){for(b in 1:floor(sqrt(c))){if(b*b+(b-1)^2==c){out<-c(out,paste(a,b));break}}}};cat(if(length(out))paste(out,collapse="\n") else "No Solution","\n")

File: PAT_Basic_Level_by_lua/1103 缘分数/1103 缘分数.lua
-- 题目编号: 1103 缘分数
-- 实现原理: 在区间[m,n]中枚举a，根据立方差公式推导出条件3a²-3a+1应为完全平方数（记为c²），且2c-1也应为完全平方数（记为r²）且r为奇数，满足条件的a即为缘分数，输出a与其对应的另一个数(r+1)/2。
local m, n = io.read("*n"), io.read("*n")  -- 读取区间范围[m, n]
local any = false                          -- 标记是否找到缘分数

-- 枚举区间内的每个数a
for a = m, n do
    -- 根据立方差公式推导出的中间量：q = 3a² - 3a + 1
    local q = 3 * a * a - 3 * a + 1
    -- 计算q的平方根并取整（加0.5避免浮点误差）
    local c = math.floor(math.sqrt(q) + .5)
    
    -- 判断q是否为完全平方数（即c² == q）
    if c * c == q then
        -- 进一步判断2c-1是否为完全平方数
        local r = math.floor(math.sqrt(2 * c - 1) + .5)
        -- 同时要求r必须为奇数
        if r * r == 2 * c - 1 and r % 2 == 1 then
            -- 输出缘分数a和对应的另一个数(r+1)/2
            print(a .. " " .. math.floor((r + 1) / 2))
            any = true
        end
    end
end

-- 如果没有找到任何缘分数，输出提示
if not any then
    print("No Solution")
end
